bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs

```diff
@@ -1,27 +1,135 @@
+    private readonly IServiceAccountRepository _serviceAccountRepository;
 
-    public DeleteAccessPolicyCommand(IAccessPolicyRepository accessPolicyRepository)
+    public DeleteAccessPolicyCommand(
+        IAccessPolicyRepository accessPolicyRepository,
+        ICurrentContext currentContext,
+        IProjectRepository projectRepository,
+        IServiceAccountRepository serviceAccountRepository)
     {
+        _projectRepository = projectRepository;
+        _serviceAccountRepository = serviceAccountRepository;
         _accessPolicyRepository = accessPolicyRepository;
+        _currentContext = currentContext;
     }
 
 
-    public async Task DeleteAsync(Guid id)
+    public async Task DeleteAsync(Guid id, Guid userId)
     {
         var accessPolicy = await _accessPolicyRepository.GetByIdAsync(id);
         if (accessPolicy == null)
         {
             throw new NotFoundException();
         }
 
+        switch (accessPolicy)
+        {
+            case UserProjectAccessPolicy ap:
+                if (ap.GrantedProjectId == null)
+                {
+                    throw new BadRequestException();
+                }
+
+                var project = await _projectRepository.GetByIdAsync(ap.GrantedProjectId.Value);
+                await DeleteProjectGrantAsync(ap.Id, userId, project.OrganizationId, project.Id);
+                break;
+            case GroupProjectAccessPolicy ap:
+                await DeleteProjectGrantAsync(ap.Id, userId, ap.Group?.OrganizationId, ap.GrantedProjectId);
+                break;
+            case ServiceAccountProjectAccessPolicy ap:
+                await DeleteProjectGrantAsync(ap.Id, userId, ap.ServiceAccount?.OrganizationId, ap.GrantedProjectId);
+                break;
+            case UserServiceAccountAccessPolicy ap:
+                await DeleteServiceAccountGrantAsync(ap.Id, userId, ap.GrantedServiceAccountId);
+                break;
+            case GroupServiceAccountAccessPolicy ap:
+                await DeleteServiceAccountGrantAsync(ap.Id, userId, ap.GrantedServiceAccountId);
+                break;
+            default:
+                throw new ArgumentException("Unsupported access policy type provided.");
+        }
+    }
+
+    private async Task DeleteProjectGrantAsync(Guid id, Guid userId, Guid? organizationId, Guid? projectId)
+    {
+        if (organizationId == null || projectId == null)
+        {
+            throw new BadRequestException();
+        }
+
+        await CheckPermissionsAsync(organizationId.Value, userId, projectIdToCheck: projectId.Value);
+        await _accessPolicyRepository.DeleteAsync(id);
+    }
+
+    private async Task DeleteServiceAccountGrantAsync(Guid id, Guid userId, Guid? serviceAccountId)
+    {
+        if (serviceAccountId == null)
+        {
+            throw new BadRequestException();
+        }
+
+        var serviceAccount = await _serviceAccountRepository.GetByIdAsync(serviceAccountId.Value);
```

Comment:
Could we just include the service account in the access policy directly? Should let us simplify this as well. We could move the actual access  check into the CheckPermissions, and move out the delete async into the regular method.

Reply:
Resolved in https://github.com/bitwarden/server/pull/2628/commits/cd92ea867e0c629cdebeccbb8a3d9499053cef04